ios/engine/KMEI/KeymanEngine/Classes/KeymanInputViewController.swift

```diff
@@ -0,0 +1,308 @@
+    let sRange = fragment.range(of: "+s=")!
+
+    let dn = Int(fragment[dnRange.upperBound..<sRange.lowerBound])!
+    let s = fragment[sRange.upperBound...]
+
+    let hexStrings: [String]
+    if !s.isEmpty {
+      hexStrings = s.components(separatedBy: ",")
+    } else {
+      hexStrings = []
+    }
+
+    let codeUnits = hexStrings.map { hexString -> UInt16 in
+      let scanner = Scanner(string: hexString)
+      var codeUnit: UInt32 = 0
+      scanner.scanHexInt32(&codeUnit)
+      return UInt16(codeUnit)
+    }
+
+    let text = String(utf16CodeUnits: codeUnits, count: codeUnits.count)
+
+    if dn <= 0 {
+      textDocumentProxy.insertText(text)
+      return
+    }
+
+    for _ in 0..<dn {
+      let oldContext = textDocumentProxy.documentContextBeforeInput ?? ""
+      textDocumentProxy.deleteBackward()
+      let newContext = textDocumentProxy.documentContextBeforeInput ?? ""
+      let unitsDeleted = oldContext.utf16.count - newContext.utf16.count
+      if unitsDeleted > 1 {
+        if !KeymanInputViewController.isSurrogate(oldContext.utf16.last!) {
+          let lowerIndex = oldContext.utf16.index(oldContext.utf16.startIndex,
+                                                  offsetBy: newContext.utf16.count)
+          let upperIndex = oldContext.utf16.index(lowerIndex, offsetBy: unitsDeleted - 1)
+          textDocumentProxy.insertText(String(oldContext[lowerIndex..<upperIndex]))
+        }
+      }
+    }
+
+    if !text.isEmpty {
+      textDocumentProxy.insertText(text)
+    }
+  }
+
+  @objc public func updatedFragment(_ fragment: String) {
+    if fragment.isEmpty {
+      return
+    }
+
+    if fragment.contains("insertText") {
+      processInsertText(fragment)
+    } else if fragment.contains("menuKeyUp") {
+      if Manager.shared.isKeyboardMenuVisible {
+        return
+      }
+
+      switch globeKeyTapBehaviour {
+      case .switchToNextKeyboard:
+        if let nextIndex = Manager.shared.switchToNextKeyboard(), nextIndex <= 0 {
+          advanceToNextInputMode()
+        }
+      case .switchToNextInputMethod:
+        advanceToNextInputMode()
+      case .doNothing:
+        break
+      }
+    } else if fragment.contains("showKeyboardMenu") {
+      switch menuBehaviour {
+      case .showAlways,
+           .showIfMultipleKeyboards where keyboardListCount > 1:
+        Manager.shared.showKeyboardMenu(self, closeButtonTitle: menuCloseButtonTitle)
+      case .showIfMultipleKeyboards, // keyboardListCount() <= 1
+           .showNever:
+        break
+      }
+    } else if fragment.contains("hideKeyboard") {
+      dismissKeyboard()
+    }
+  }
+
+  private func setConstraints() {
+    let topBarHeight = isTopBarEnabled ? KeymanInputViewController.topBarHeight : 0
+    let viewsDict = ["bar": topBarImageView!, "container": containerView!]
+    barHeightConstraints = NSLayoutConstraint.constraints(
+      withVisualFormat: "V:[bar(\(topBarHeight))]", metrics: nil, views: viewsDict)
```

Comment:
The string components of `withVirtualFormat` here don't match the originals from Obj-C.  Is the required format for Swift different?

Reply:
The format should be the same since the format is API defined rather than a language feature. Here's the [API reference](https://developer.apple.com/library/content/documentation/UserExperience/Conceptual/AutolayoutPG/VisualFormatLanguage.html).

The name of the views in the dict were changed though and the new code using Swift string interpolation so that might be why it looks different.